File: 279/perfect-squares.js
/**
 * @param {number} n
 * @return {number}
 */
var numSquares = function (n) {
    let hash = [];
    for (let i = 1; i <= n; i++) {
        hash[i] = Infinity;
    }
    hash[0] = 0;

    for (let i = 0; i <= n; i++) {
        for (let j = 1; i + j * j <= n; j++) {
            hash[i + j * j] = Math.min(hash[i + j * j], hash[i] + 1);
        }
    }
    return hash[n];
};

File: 279/PerfectSquares.scala
object PerfectSquares {
  def numSquares(n: Int): Int = {
    val hash = Array.fill(n+1)(Int.MaxValue)
    hash(0) = 0

    for (i <- 0 to n) {
      for (j <- 1 to n if i + j*j <= n) {
        hash(i+j*j) = hash(i+j*j) min (hash(i) + 1)
      }
    }
    hash(n)
  }

  def main(args: Array[String]): Unit = {
    val n = 13
    println(numSquares(n))
  }
}
